src/Stratis.Bitcoin/Consensus/ChainedHeaderTree.cs

```diff
@@ -0,0 +1,536 @@
+        private readonly Dictionary<uint256, ChainedHeader> chainedHeadersByHash;
+
+        internal Dictionary<uint256, ChainedHeader> GetChainedHeadersByHash => this.chainedHeadersByHash;
+        internal Dictionary<uint256, HashSet<int>> GetPeerIdsByTipHash => this.peerIdsByTipHash;
+        internal Dictionary<int, uint256> GetPeerTipsByPeerId => this.peerTipsByPeerId;
+
+        public ChainedHeaderTree(
+            Network network, 
+            ILoggerFactory loggerFactory, 
+            IChainedHeaderValidator chainedHeaderValidator, 
+            ICheckpoints checkpoints, 
+            IChainState chainState, 
+            ConsensusSettings consensusSettings)
+        {
+            this.network = network;
+            this.chainedHeaderValidator = chainedHeaderValidator;
+            this.checkpoints = checkpoints;
+            this.chainState = chainState;
+            this.consensusSettings = consensusSettings;
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+
+            this.peerTipsByPeerId = new Dictionary<int, uint256>();
+            this.peerIdsByTipHash = new Dictionary<uint256, HashSet<int>>();
+            this.chainedHeadersByHash = new Dictionary<uint256, ChainedHeader>();
+        }
+
+        /// <summary>
+        /// Initialize the tree with a <see cref="ChainedHeader"/> that is the tip of our best known consensus chain.
+        /// </summary>
+        /// <param name="consensusTip">The consensus tip.</param>
+        public void Initialize(ChainedHeader consensusTip)
+        {
+            this.logger.LogTrace("({0}:'{1}')", nameof(consensusTip), consensusTip);
+
+            ChainedHeader current = consensusTip;
+            while (current.Previous != null)
+            {
+                current.Previous.Next.Add(current);
+                this.chainedHeadersByHash.Add(current.HashBlock, current);
+                current = current.Previous;
+            }
+
+            if (current.HashBlock != this.network.GenesisHash)
+            {
+                this.logger.LogTrace("(-)[INVALID_NETWORK]");
+                throw new ConsensusException("INVALID_NETWORK");
+            }
+
+            this.SetPeerTip(LocalPeerId, consensusTip.HashBlock);
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <summary>
+        /// A new list of headers are presented from a given peer, the headers will try to be connected to the chain.
+        /// Headers that are interesting (i.e may extend our consensus tip) will marked to download there full blocks.
+        /// </summary>
+        /// <remarks>
+        /// The headers are assumed to be consecutive in order. 
+        /// </remarks>
+        /// <param name="networkPeerId">The network id that is presenting this headers.</param>
+        /// <param name="headers">The list of headers to connect to the chain tree.</param>
+        /// <returns>Indicators of what blocks need to be downloaded.</returns>
```

Comment:
Information about which blocks need to be downloaded together with information about which input headers were processed. Only headers that we can validate will be processed. The rest of the headers will be submitted later again for processing.